File: CPP/binary_tree_zigzag_level_order_traversal.cpp
class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        vector<vector<int>> result;
        queue<TreeNode *> q;
        int curr_level_node_count = 0;
        int next_level_node_count = 0;
        if (root != NULL) {
            curr_level_node_count = 1;
            q.push(root);
        }
        while (!q.empty()) {
            next_level_node_count = 0;
            vector<int> curr_level_value;
            while (curr_level_node_count--) {
                root = q.front();
                q.pop();
                if (root->left) {
                    q.push(root->left);
                    next_level_node_count++;
                }
                if (root->right) {
                    q.push(root->right);
                    next_level_node_count++;
                }
                curr_level_value.push_back(root->val);
            }
            result.push_back(curr_level_value);
            curr_level_node_count = next_level_node_count;
        }
        for (int i = 0; i < result.size(); i++) {
            if (i % 2 == 1) {
                reverse(result[i].begin(), result[i].end());
            }
        }
        
        return result;
    }
};

File: JAVA/BinaryTreeZigzagLevelOrderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// recursive
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        zigzagHelper(root, res, 0);
        return res;
    }
    
    private void zigzagHelper(TreeNode root, List<List<Integer>> res, int level) {
        if (root == null) return;
        if (res.size() == level) res.add(new ArrayList<Integer>());
        if (level % 2 == 0) res.get(level).add(root.val);
        else res.get(level).add(0, root.val);
        zigzagHelper(root.left, res, level + 1);
        zigzagHelper(root.right, res, level + 1);
    }
}

public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new LinkedList<List<Integer>>();
        if (root == null)
            return result;
        
        List<List<TreeNode>> levelNodes = new ArrayList<List<TreeNode>>();
        List<TreeNode> level = new ArrayList<TreeNode>();
        level.add(root);
        levelNodes.add(level);
        boolean zig = true;
        int i = 0;
        while (i < levelNodes.size()) {
            List<Integer> levelRst = new LinkedList<Integer>();
            level = new ArrayList<TreeNode>();
            
            List<TreeNode> traversal = levelNodes.get(i);
            for (TreeNode node : traversal) {
                levelRst.add(node.val);
            }
            Collections.reverse(traversal);
            for (TreeNode node : traversal) {
                
                if (zig) {
                    if (node.right != null)
                        level.add(node.right);
                    if (node.left != null)
                        level.add(node.left);
                } else {
                    if (node.left != null)
                        level.add(node.left);
                    if (node.right != null)
                        level.add(node.right);
                }
            }
            result.add(levelRst);
            if (!level.isEmpty())
                levelNodes.add(level);
            zig = zig ? false : true;
            i++;
        }
        return result;
    }
}